### PR TITLE
Adding Capability to configure the agent for machine group (RI rmmaster2 to master)

### PR DESCRIPTION
**Pull request by chshrikh**

### CODE REVIEW

File: src/Agent.Listener/Configuration/ConfigurationProvider.cs [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+
+        public void UpdateAgentSetting(AgentSettings settings)
+        {
+            settings.MachineGroupName = _machineGroupName;
+            settings.ProjectName = _projectName;
+        }
+
+        private async Task<int> GetPoolIdAsync(string projectName, string machineGroupName)
+        {
+            int poolId = 0;
```
Comment (TingluoHuang) [inline src/Agent.Listener/Configuration/ConfigurationProvider.cs]: not need this local var, return machiengroup.pool.id

File: src/Agent.Listener/Configuration/ConfigurationProvider.cs [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+            Trace.Info("Found machine group {0} with id {1}", machineGroupName, machineGroupId);
+            poolId = machineGroup.Pool.Id;
+            Trace.Info("Found poolId {0} for machine group {1}", poolId, machineGroupName);
+
+            return poolId;
+        }
+
+        private void ThrowExceptionForOnPremUrl()
+        {
+            throw new Exception(StringUtil.Loc("UrlValidationFailedForOnPremTfs"));
```
Comment (TingluoHuang) [inline src/Agent.Listener/Configuration/ConfigurationProvider.cs]: throw these inline, you only use it three times, but adding 10 lines of code.

File: src/Agent.Listener/Configuration/ConfigurationProvider.cs [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+        private IMachineGroupServer _machineGroupServer = null;
+
+        public string ConfigurationProviderType
+            => Constants.Agent.AgentConfigurationProvider.DeploymentAgentConfiguration;
+
+        public override void Initialize(IHostContext hostContext)
+        {
+            base.Initialize(hostContext);
+            _term = hostContext.GetService<ITerminal>();
+            _agentServer = HostContext.GetService<IAgentServer>();
```
Comment (TingluoHuang) [inline src/Agent.Listener/Configuration/ConfigurationProvider.cs]: _machinegroupserver=hostcontext.getservice<>()

**Comment by TingluoHuang** [comment]
have you try config a build/releases agent with your changes?

File: src/Agent.Listener/Configuration/ConfigurationManager.cs [hunk 4]
```diff
@@ -102,26 +102,37 @@ public AgentSettings LoadSettings()
                     throw new NotSupportedException();
             }
 
+            // Create the configuration provider as per agent type.....
+            string agentType = command.MachineGroup
```
Comment (chshrikh) [inline src/Agent.Listener/Configuration/ConfigurationManager.cs]: yes

File: src/Agent.Listener/Configuration/ConfigurationManager.cs [hunk 5]
```diff
@@ -381,14 +384,23 @@ public AgentSettings LoadSettings()
                     await agentSvr.ConnectAsync(conn);
                     Trace.Info("Connect complete.");
 
+                    Trace.Info("Agent configured for machineGroup : {0}", settings.MachineGroup.ToString());
```
Comment (chshrikh) [inline src/Agent.Listener/Configuration/ConfigurationManager.cs]: no I checked that it come as false I have varified that

**Comment by chshrikh** [comment]
@ have you try config a build/releases agent with your changes? - 

Yes, it works fine

File: src/Agent.Listener/Configuration/ConfigurationProvider.cs (hunk 3 above)
Comment (chshrikh) [inline src/Agent.Listener/Configuration/ConfigurationProvider.cs]: taken care

File: src/Agent.Listener/Configuration/ConfigurationProvider.cs (hunk 1 above)
Comment (chshrikh) [inline src/Agent.Listener/Configuration/ConfigurationProvider.cs]: taken care

File: src/Agent.Listener/Configuration/ConfigurationProvider.cs [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+            {
+                throw new TaskAgentPoolNotFoundException(StringUtil.Loc("PoolNotFound", poolName));
+            }
+            else
+            {
+                Trace.Info("Found pool {0} with id {1}", poolName, agentPool.Id);
+                return agentPool.Id;
+            }
+        }
+
```
Comment (chshrikh) [inline src/Agent.Listener/Configuration/ConfigurationProvider.cs]: taken care :)

File: src/Agent.Listener/CommandSettings.cs [hunk 7]
```diff
@@ -164,15 +168,35 @@ public string GetToken()
                 validator: Validators.NonEmptyValidator);
         }
 
-        public string GetUrl()
+        public string GetUrl(bool isMachineGroup = false)
```
Comment (chshrikh) [inline src/Agent.Listener/CommandSettings.cs]: sure, taken care

File: src/Agent.Listener/Configuration/ConfigurationProvider.cs (hunk 2 above)
Comment (chshrikh) [inline src/Agent.Listener/Configuration/ConfigurationProvider.cs]: sure taken care

File: src/Agent.Listener/Configuration/ConfigurationManager.cs [hunk 8]
```diff
@@ -144,22 +156,9 @@ public AgentSettings LoadSettings()
...
-                    break;
-                }
-                catch (Exception e) when (!command.Unattended)
-                {
-                    _term.WriteError(e);
-                    _term.WriteError(StringUtil.Loc("FailedToFindPool"));
-                }
-            }
 
+            poolId = await agentProvider.GetPoolId(command);
```
Comment (chshrikh) [inline src/Agent.Listener/Configuration/ConfigurationManager.cs]: loop is in provider

**Pull request merged**